package rest_errors

import (
	"errors"
	"net/http"
)

// RestErr struct describes the common API error structure
type RestErr struct {
	Message string        `json:"message"`
	Status  int           `json:"status"`
	Error   string        `json:"error"`
	Causes  []interface{} `json:"causes"`
}

// NewError returns the logger errors
func NewError(msg string) error {
	return errors.New(msg)
}

// NewBadRequestError function returns the "bad request" error
func NewBadRequestError(message string) *RestErr {
	return &RestErr{
		Message: message,
		Status:  http.StatusBadRequest,
		Error:   "bad_request",
	}
}

// NewNotFoundError function returns the "not found request" error
func NewNotFoundError(message string) *RestErr {
	return &RestErr{
		Message: message,
		Status:  http.StatusNotFound,
		Error:   "not_found",
	}
}

// NewInternalServerError function returns the "internal server error" error
func NewInternalServerError(message string, err error) *RestErr {
	result := &RestErr{
		Message: message,
		Status:  http.StatusInternalServerError,
		Error:   "internal_server_error",
	}
	if err != nil {
		result.Causes = append(result.Causes, err.Error())
	}
	return result
}

// NewNotAuthorizedError function returns the "unauthorized error" error
func NewNotAuthorizedError(message string) *RestErr {
	result := &RestErr{
		Message: "enable to retrieve user information from givet access token",
		Status:  http.StatusUnauthorized,
		Error:   "unauthorized",
	}
	return result
}
